common/main/java/com/couchbase/lite/internal/listener/ChangeNotifier.java

```diff
@@ -27,7 +27,27 @@
 import com.couchbase.lite.ListenerToken;
 import com.couchbase.lite.internal.utils.Fn;
 
-
+// Used for Document and Collection (and MessageEndpoint) change notification.
+// When used for a collection, we register a single callback with Core.
+// That single callback is delegated to each of the listeners registered
+// with this notifier.
+// Document notifiers are similar: They are created, one per document and
+// registered with Core.  A single callback is delegated to all observers
+// observing the document.
+// The token created here, and returned to the client code, dominates very
+// little memory.  Notifiers, however, hold references to Core companion
+// objects.  When the last listener is removed (removeListener returns true)
+// the notifier should be closed and the companion objects freed.
+// This moots the concerning scenario in which client code simply drops
+// the token on the floor.  If it is never released, it and the notifier
+// that contains it will not be released: the Core objects will be retained.
+// Having the tokens free themselves in their finalizers is not a good approach.
+// If you do that, as soon as the token is GCed, the associated listener will
+// stop listening, surely surprising to the client code.
+// On the other hand, once the replicator is freed, all of its notifiers are
```

Comment:
Assume it means that when the objects that keep the notifiers, not just the replicator, are freed, the underlining Core's Observer will be freed?

Reply:
No.  In order to free the Core observer, one of two things must happen: either a call to token.remove() or the Replicator object becomes unreachable (in the Java sense)